libsplinter/src/admin/service/shared.rs

```diff
@@ -390,89 +391,143 @@ impl AdminServiceShared {
+                                let mut member_ready = MemberReady::new();
+                                member_ready.set_circuit_id(circuit_id.to_string());
+                                member_ready.set_member_node_id(self.node_id.clone());
+                                let mut msg = AdminMessage::new();
+                                msg.set_message_type(AdminMessage_Type::MEMBER_READY);
+                                msg.set_member_ready(member_ready);
+
+                                let envelope_bytes =
+                                    msg.write_to_bytes().map_err(MarshallingError::from)?;
+                                for member in circuit.members().iter() {
+                                    if member != &self.node_id {
+                                        network_sender
+                                            .send(&admin_service_id(member), &envelope_bytes)?;
+                                    }
+                                }
+                            }
 
-                        self.routing_table_writer
-                            .add_circuit(
-                                circuit.circuit_id().to_string(),
-                                routing_circuit,
-                                routing_members,
-                            )
-                            .map_err(|_| {
+                            // add circuit as pending initialization
+                            self.add_uninitialized_circuit(circuit_proposal.clone())?;
+                        } else if status == Circuit_CircuitStatus::DISBANDED {
+                            // Circuit has been disbanded: all associated services will be shut
+                            // down, the circuit removed from the routing table, and peer refs
+                            // for this circuit will be removed.
+                            self.shutdown_services(circuit_proposal.get_circuit_proposal())?;
+                            // Removing the circuit from the routing table
+                            self.routing_table_writer
+                                .remove_circuit(circuit_id)
+                                .map_err(|_| {
+                                    AdminSharedError::SplinterStateError(format!(
+                                        "Unable to remove circuit from routing table: {}",
+                                        circuit_id
+                                    ))
+                                })?;
+                            // Removing the circuit's peer refs
+                            for member in circuit_proposal.get_circuit_proposal().get_members() {
+                                self.remove_peer_ref(member.get_node_id());
+                            }
+                            // Removing the corresponding `active` circuit from the admin store
+                            self.admin_store.remove_circuit(circuit_id).map_err(|_| {
```

Comment:
We should use update_circuit here instead of remove/upgrade to keep the operation atomic. This will require you get the circuit from the proposal here instead of relying on the admin service store to do the conversion.